scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,620 @@
+            - ("greater", lb): for a constraint of the type:
+                f[i] >= lb[i]
+            - ("less", ub): for a constraint of the type:
+                f[i] <= ub[i]
+            - ("equals", c): for a constraint of the type:
+                f[i] == c[i]
+
+        where ``lb``,  ``ub`` and ``c`` are (m,) ndarrays.
+    Returns
+    -------
+    eq : array_like
+        A vector indicating equality constraints.
+            len(eq) = number of equality constraints
+    ineq : array_like
+        A vector indicating inequality constraints.
+            len(ineq) = number of inequality constraints
+    val_eq : array_like
+        Equality constraint right hand side:
+            f[eq[i]] = val_eq[i]
+    val_ineq : array_like
+        Inequality constraint right hand side:
+            sign[i]*(f[ineq[i]] - val_ineq[i]) <= 0
+    sign : array_like
+        Sign of inequality constraints.
+    """
+
+    if kind[0] == "equals":
+        # Read values from input structure
+        c = np.asarray(kind[1], dtype=float)
+        # Set returns
+        eq = np.arange(len(c), dtype=int)
+        ineq = np.empty(0)
+        val_eq = np.asarray(c)
+        val_ineq = np.empty(0)
+        sign = np.empty(0)
+        fun_len = len(c)
+    elif kind[0] in ("greater", "less", "interval"):
+        # Constraint type
+        if kind[0] == "greater":
+            lb = np.asarray(kind[1], dtype=float)
+            ub = np.full_like(lb, np.inf, dtype=float)
+        elif kind[0] == "less":
+            ub = np.asarray(kind[1], dtype=float)
+            lb = np.full_like(ub, -np.inf, dtype=float)
+        elif kind[0] == "interval":
+            lb = np.asarray(kind[1], dtype=float)
+            ub = np.asarray(kind[2], dtype=float)
+            if len(lb) != len(ub):
+                raise ValueError("Mismatching 'ub' and 'lb' dimensions.")
+            if (lb > ub).any():
+                raise ValueError("lb[i] > ub[i]")
+        # Set auxiliar values
+        arange = np.arange(len(lb), dtype=int)
+        ones = np.ones(len(lb))
+        lb_isinf = np.isinf(lb)
+        ub_isinf = np.isinf(ub)
+        eq_list = (lb == ub) & ~lb_isinf & ~ub_isinf
+        # Set returns
+        eq = arange[eq_list]
+        val_eq = lb[eq_list]
+        ineq = np.hstack((arange[~eq_list & ~lb_isinf],
+                          arange[~eq_list & ~ub_isinf]))
+        val_ineq = np.hstack((lb[~eq_list & ~lb_isinf],
+                              ub[~eq_list & ~ub_isinf]))
+        sign = np.hstack((-ones[~eq_list & ~lb_isinf],
+                          ones[~eq_list & ~ub_isinf]))
+        fun_len = len(lb)
+    else:
+        raise ValueError("kind '%s' not available." % kind[0])
+
+    return eq, ineq, val_eq, val_ineq, sign, fun_len
+
+
+def reinforce_box_constraints(constr, x,
+                              relative_tolerance=0.01,
+                              absolute_tolerance=0.01):
+    """Reinforce box constraints"""
```

Comment:
This is for the case when the user simultaneously set ``enforce_feasibility=True`` and selects an initial point ``x0`` that violate bounds constraints. In this case I use this function to reinforce bound constraints and send a warning to the user.